app/models/reviewer_access.rb

```diff
@@ -0,0 +1,82 @@
+# manage anonymous access to private studies via access code & pin
+# will generate a short-lived session for reviewers with valid credentials
+class ReviewerAccess
+  include Mongoid::Document
+  include Mongoid::Timestamps
+  include ActiveSupport::SecurityUtils
+
+  field :access_code, type: String # UUID
+  field :pin, type: String # 6 character string
+  field :expires_at, type: Date, default: 2.months.from_now.to_date
+
+  PIN_LENGTH = 10
+  # regex to strip all non-alphanumeric characters from a user-supplied pin value
+  PIN_SANITIZER = /(\W|_)/.freeze
+
+  belongs_to :study
+  has_many :reviewer_access_sessions, dependent: :delete_all do
+    def by_session_key(session_key)
+      find_by(session_key: session_key)
```

Comment:
I _think_ we should validate `session_key` as well, unless `find_by` does it automatically.

Reply:
It does minimally... `find_by` and other parameterized query helpers will escape any unsafe characters to prevent injection, but it won't validate the format or length.  I can easily add a regex that will ensure that the provided key matches the pattern for a UUID.

Reply:
259d81c

Reply:
Sounds good, either that or maybe there's a ready-to-go `is_uuid` method. I was wary of introducing other vulns via regex itself 😄

Reply:
There is not, unfortunately.  I looked for a while.